src/System.Buffers/src/System/Buffers/ArrayPool.cs

```diff
@@ -0,0 +1,97 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace System.Buffers
+{
+    /// <summary>
+    /// This class provides resource pooling for arrays of any type. This is can increase performance
+    /// of certain applications where lots of arrays are created and destroyed in rapid succession.
+    /// This class is thread-safe.
+    /// </summary>
+    public abstract class ArrayPool<T>
+    {
+        /// <summary>The default number of arrays that are available for rent.</summary>
+        public const int DefaultNumberOfArraysPerBucket = 50;
+        /// <summary>The default length of each Rent'able array; equal to 1MB in bytes</summary>
+        public const int DefaultArrayLength = 1024000;
+
+        private static ArrayPool<T> _sharedInstance = null;
+
+        /// <summary>
+        /// Retrieves a shared instance pool that is thread-safe and can be used by 
+        /// multiple components of the same system.
+        /// </summary>
+        public static ArrayPool<T> Shared
+        {
+            get
+            {
+                if (Volatile.Read(ref _sharedInstance) == null)
+                {
+                    Interlocked.CompareExchange(ref _sharedInstance, new DefaultArrayPool<T>(DefaultArrayLength, DefaultNumberOfArraysPerBucket), null);
+                }
+
+                return _sharedInstance;
+            }
+        }
+
+        /// <summary>
+        /// Creates a new ArrayPool instance of the given type.
+        /// </summary>
+        /// <param name="maxArrayLength">The maximum length of any Rent request.</param>
+        /// <param name-"numberOfArrays">The maximum number of arrays that will be rented.</param>
+        /// <returns>Returns a new ArrayPool<T> instance with the specified configuration options</return>
+        public static ArrayPool<T> Create(int maxArrayLength = DefaultArrayLength, int numberOfArrays = DefaultNumberOfArraysPerBucket)
+        {
+            return new DefaultArrayPool<T>(maxArrayLength, numberOfArrays);
+        }
+
+        /// <summary>
+        /// Gets the number of buffers that have been Rented and not yet Returned. This count includes
+        /// any buffers that have been allocated on-demand.
+        /// </summary>
+        public abstract int RentedBufferCount { get; }
+
+        /// <summary>
+        /// Gets the total number of allocated buffers available for Rent. This count includes any outstanding
+        /// buffers that have been allocated on-demand.
+        /// </summary>
+        public abstract int TotalBufferCount { get; }
+
+        /// <summary>
+        /// Retrieves a buffer from the pool that is at least the requested length. This buffer is loaned
+        /// to the caller and the caller must call Return when finished with the buffer.
+        /// </summary>
```

Comment:
I think there will be other policies for different scenarios, so I tried to purposely leave implementation details out of the comments and contract. I think once we figure out the "overriding the shared pool creation" issue, the need for async APIs or TryRent goes away since consumers that need that can create their own pool that wraps the default pool. What do you guys think?